OG-134 Not ready yet is not informative

forshtat opened the pull request:
Also, some progress on OG-139 as logs are improved throughout the
relaying code (WIP)

1. Remove 'StateError' as a way to bail out when not ready.
2. Create 'AmountRequired' to maybe save on suplicate logic piece
(meeting some ETH value criteria)

drortirosh reviewed (commented):
Overall, very nice. and it actually works..
I thought that `chalk` will break logs (non-interactive), but it works ok (that is, colors are silently removed)
I ran a test with "ganache -b 1", and console for server operations.

- "not satisfied" - not a good name. should be short and single word. maybe `ok/wrong` ?
- "latest block" - overly verbose. maybe set at `trace` level (lower than `debug`), so will not appear by default.
- Stake: required:0 eth, instead of reading from hub.
- if I stake for too less or delay too low, it behaves funny and completely counter-intuitive: repeated reverted TXs in the log with no obvious reason.
- no need to show "worker balance" in the "summary" page - its a completely internal issue (can show the state change of it)
- `loglevel` - can't set from config
- `debug` - flag exists, but does nothing
- `devMode` - hard-coded on 2 values: true (0.1sec) and false (10-sec). instead, its should only be a default for a new config:
- `checkInterval` - interval in seconds to check blockchain. defaults to 10 in devMode=false and 0.1 in devMode=true.

drortirosh commented on src/relayserver/RegistrationManager.ts:
```diff
@@ -76,6 +63,32 @@ export class RegistrationManager {
+  set isHubAuthorized (newValue: boolean) {
+    const oldValue = this._isHubAuthorized
+    this._isHubAuthorized = newValue
+    if (newValue !== oldValue) {
```
this is how log looks if I stakeForRelay with too low values: I get the following repeatedly on each block:
(there are 3 problems here: 1. the "stake:satisified" is wrong (I put only 0.1eth), 2. the unstakeDelay is wrong (which you can't even see)
But the most important: the logged TX is meaningless: you report TXid, from,to, and if you really want to go extra level is to report the method-sig
```
Not registered yet. Prerequisites:
Balance        | satisfied      | actual:       1.0000 ETH | required:       0.1000 ETH
Stake          | satisfied      | actual:       0.0100 ETH | required:       0.0000 ETH
Hub authorized | not satisfied 
Stake locked   | satisfied     
Manager        | 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
Worker         | 0x4e185eb8ae0a7a404b1bb822db2cc5a05d755292
Owner          | unknown

Got staked for the first time

Not registered yet. Prerequisites:
Balance        | satisfied      | actual:       1.0000 ETH | required:       0.1000 ETH
Stake          | satisfied      | actual:       0.0100 ETH | required:       0.0000 ETH
Hub authorized | not satisfied 
Stake locked   | satisfied     
Manager        | 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
Worker         | 0x4e185eb8ae0a7a404b1bb822db2cc5a05d755292
Owner          | 0x25F974222f4DDE02b30c1ebd63C63aE9D5A1bB27

resending unconfirmed transactions
failed to fetch receipt for tx 0xf3881cfd15f3291b267e8417e3637a528941097de6eee5fce51f7a6ce559de48
failed to fetch receipt for tx 0x1a93ce0fb5150dece1fc33ce9fab2092f5bd8f365673cde13f272a9eb1d06e3c
failed to fetch receipt for tx 0x8a0c0461712f336757bbc87c037a960cb1a53e8b781bce691307921e0739c56f
txToSign Transaction {
  _common: Common {
    _chainParams: {
      name: 'mainnet',
      chainId: 1337,
      networkId: 1601029322839,
      comment: 'The Ethereum main chain',
      url: 'https://ethstats.net/',
      genesis: [Object],
      hardforks: [Array],
      bootstrapNodes: [Array]
    },
    _hardfork: 'istanbul',
    _supportedHardforks: []
  },
  raw: [
    <Buffer >,
    <Buffer 05 96 82 f0 00>,
    <Buffer 01 31 d3>,
    <Buffer 25 4d ff cd 32 77 c0 b1 66 0f 6d 42 ef bb 75 4e da ba bc 2b>,
    <Buffer >,
    <Buffer c2 da 07 86 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 50 more bytes>,
    <Buffer >,
    <Buffer >,
    <Buffer >
  ],
  _fields: [
    'nonce',    'gasPrice',
    'gasLimit', 'to',
    'value',    'data',
    'v',        'r',
    's'
  ],
  toJSON: [Function (anonymous)],
  serialize: [Function: serialize],
  nonce: [Getter/Setter],
  gasPrice: [Getter/Setter],
  gasLimit: [Getter/Setter],
  to: [Getter/Setter],
  value: [Getter/Setter],
  data: [Getter/Setter],
  v: [Getter/Setter],
  r: [Getter/Setter],
  s: [Getter/Setter],
  from: [Getter]
}
resending tx with nonce <Buffer > from 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
account nonce 0


```

drortirosh commented:
this is how log looks if I stakeForRelay with too low values: I get the following repeatedly on each block:
(there are 3 problems here: 1. the "stake:satisified" is wrong (I put only 0.1eth), 2. the unstakeDelay is wrong (which you can't even see)
But the most important: the logged TX is meaningless: you report TXid, from,to, and if you really want to go extra level is to report the method-sig
```
Not registered yet. Prerequisites:
Balance        | satisfied      | actual:       1.0000 ETH | required:       0.1000 ETH
Stake          | satisfied      | actual:       0.0100 ETH | required:       0.0000 ETH
Hub authorized | not satisfied 
Stake locked   | satisfied     
Manager        | 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
Worker         | 0x4e185eb8ae0a7a404b1bb822db2cc5a05d755292
Owner          | unknown

Got staked for the first time

Not registered yet. Prerequisites:
Balance        | satisfied      | actual:       1.0000 ETH | required:       0.1000 ETH
Stake          | satisfied      | actual:       0.0100 ETH | required:       0.0000 ETH
Hub authorized | not satisfied 
Stake locked   | satisfied     
Manager        | 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
Worker         | 0x4e185eb8ae0a7a404b1bb822db2cc5a05d755292
Owner          | 0x25F974222f4DDE02b30c1ebd63C63aE9D5A1bB27

resending unconfirmed transactions
failed to fetch receipt for tx 0xf3881cfd15f3291b267e8417e3637a528941097de6eee5fce51f7a6ce559de48
failed to fetch receipt for tx 0x1a93ce0fb5150dece1fc33ce9fab2092f5bd8f365673cde13f272a9eb1d06e3c
failed to fetch receipt for tx 0x8a0c0461712f336757bbc87c037a960cb1a53e8b781bce691307921e0739c56f
txToSign Transaction {
  _common: Common {
    _chainParams: {
      name: 'mainnet',
      chainId: 1337,
      networkId: 1601029322839,
      comment: 'The Ethereum main chain',
      url: 'https://ethstats.net/',
      genesis: [Object],
      hardforks: [Array],
      bootstrapNodes: [Array]
    },
    _hardfork: 'istanbul',
    _supportedHardforks: []
  },
  raw: [
    <Buffer >,
    <Buffer 05 96 82 f0 00>,
    <Buffer 01 31 d3>,
    <Buffer 25 4d ff cd 32 77 c0 b1 66 0f 6d 42 ef bb 75 4e da ba bc 2b>,
    <Buffer >,
    <Buffer c2 da 07 86 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 50 more bytes>,
    <Buffer >,
    <Buffer >,
    <Buffer >
  ],
  _fields: [
    'nonce',    'gasPrice',
    'gasLimit', 'to',
    'value',    'data',
    'v',        'r',
    's'
  ],
  toJSON: [Function (anonymous)],
  serialize: [Function: serialize],
  nonce: [Getter/Setter],
  gasPrice: [Getter/Setter],
  gasLimit: [Getter/Setter],
  to: [Getter/Setter],
  value: [Getter/Setter],
  data: [Getter/Setter],
  v: [Getter/Setter],
  r: [Getter/Setter],
  s: [Getter/Setter],
  from: [Getter]
}
resending tx with nonce <Buffer > from 0x3eca7d8fa4363cedfd0298fcb7889fba5b4d1651
account nonce 0


```

forshtat commented:
> Overall, very nice. and it actually works..
> (several code-specific comments below, which I wrote before running it)
> 
> I thought that `chalk` will break logs (non-interactive), but it works ok (that is, colors are silently removed)
> I ran a test with "ganache -b 1", and console for server operations.
> 
> * "not satisfied" - not a good name. should be short and single word. maybe `ok/wrong` ?
> * "latest block" - overly verbose. maybe set at `trace` level (lower than `debug`), so will not appear by default.
> * Stake: required:0 eth, instead of reading from hub.
> * if I stake for too less or delay too low, it behaves funny and completely counter-intuitive: repeated reverted TXs in the log with no obvious reason (see separate comment below)
> * no need to show "worker balance" in the "summary" page - its a completely internal issue (can show the state change of it)
> * `loglevel` - can't set from config
> * `debug` - flag exists, but does nothing
> * `devMode` - hard-coded on 2 values: true (0.1sec) and false (10-sec). instead, its should only be a default for a new config:
> * `checkInterval` - interval in seconds to check blockchain. defaults to 10 in devMode=false and 0.1 in devMode=true.

1. Fixed (good/wrong). Still printing 'not satisfied' in full sentences
2. Set to 'debug' as trace will also dump the stack trace, making logs entirely unreadable. Not printing that will harm debugging IMO. On real network should not be way too much
3. OG-262 - Read 'minimum stake' value from the RelayHub
4. OG-263 - Reverted transactions (register, add worker etc) lead to an undefined behaviour
5. That is what it is now. Unfortunately, I don't have a status page, so it is just an improved 'not registered yet' log. Worker balance is only logged once it is changed (or queried for the first time IIRC)
6. See line `RelayServer.ts:80` :`log.setLevel(this.config.logLevel)`. TBH did not test but something that should better work.
7. There is no longer flag called `debug` in `SeverConfigParams.ts`
8. Added `checkInterval`, kept `devMode` for usage in 'runServer' which I think is incorrect

drortirosh commented on dockers/jsrelay/config/gsn-relay-config.json:
```diff
@@ -3,8 +3,8 @@
+  "gasPriceFactor": 1.1,
   "ethereumNodeUrl": "http://host.docker.internal:8545",
-  "devMode": true,
+  "devMode": false,
```
remove debug, add default loglevel
